api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2557,29 +2566,102 @@ public void composeObject(
+   *        .build());
+   *
+   * // Create my-bucketname/my-objectname with user metadata by combining source object
+   * // list.
+   * minioClient.composeObject(
+   *     ComposeObjectArgs.builder()
+   *        .bucket("my-bucketname")
+   *        .object("my-objectname")
+   *        .sources(sourceObjectList)
+   *        .build());
+   *
+   * // Create my-bucketname/my-objectname with user metadata and server-side encryption
+   * // by combining source object list.
+   * minioClient.composeObject(
+   *   ComposeObjectArgs.builder()
+   *        .bucket("my-bucketname")
+   *        .object("my-objectname")
+   *        .sources(sourceObjectList)
+   *        .ssec(sse)
+   *        .build());
+   *
+   * }</pre>
+   *
+   * @param args {@link ComposeObjectArgs} object.
+   * @return {@link ObjectWriteResponse} object.
+   * @throws ErrorResponseException thrown to indicate S3 service returned an error response.
+   * @throws IllegalArgumentException throws to indicate invalid argument passed.
+   * @throws InsufficientDataException thrown to indicate not enough data available in InputStream.
+   * @throws InternalException thrown to indicate internal library error.
+   * @throws InvalidBucketNameException thrown to indicate invalid bucket name passed.
+   * @throws InvalidKeyException thrown to indicate missing of HMAC SHA-256 library.
+   * @throws InvalidResponseException thrown to indicate S3 service returned invalid or no error
+   *     response.
+   * @throws IOException thrown to indicate I/O error on S3 operation.
+   * @throws NoSuchAlgorithmException thrown to indicate missing of MD5 or SHA-256 digest library.
+   * @throws XmlParserException thrown to indicate XML parsing error.
+   */
+  public ObjectWriteResponse composeObject(ComposeObjectArgs args)
+      throws ErrorResponseException, IllegalArgumentException, InsufficientDataException,
+          InternalException, InvalidBucketNameException, InvalidKeyException,
+          InvalidResponseException, IOException, NoSuchAlgorithmException, ServerException,
+          XmlParserException {
+
+    checkWriteRequestSse(args.sse());
 
     long objectSize = 0;
     int partsCount = 0;
+    List<ComposeSource> sources = args.sources();
     for (int i = 0; i < sources.size(); i++) {
       ComposeSource src = sources.get(i);
 
       checkReadRequestSse(src.ssec());
 
-      ObjectStat stat = statObject(src.bucketName(), src.objectName(), src.ssec());
+      ObjectStat stat =
+          statObject(
+              StatObjectArgs.builder()
+                  .bucket(src.bucket())
+                  .object(src.object())
+                  .versionId(src.versionId())
+                  .ssec(src.ssec())
+                  .build());
+
       src.buildHeaders(stat.length(), stat.etag());
 
-      if (i != 0 && src.headers().containsKey("x-amz-meta-x-amz-key")) {
+      if (i != 0 && src.extraHeaders().containsKey("x-amz-meta-x-amz-key")) {
```

Comment:
You would need to use `headers()` not `extraHeaders()`. `headers` member data is updated in `ComposeSource.buildHeaders()`